BaragonService/src/main/java/com/hubspot/baragon/service/elb/ApplicationLoadBalancer.java

```diff
@@ -29,23 +62,401 @@ public ApplicationLoadBalancer(Optional<ElbConfiguration> configuration,
+            TargetGroup targetGroup = maybeTargetGroup.get();
+
+            LOG.debug("Registering new instances for group {}", baragonGroup.getName());
+            guaranteeRegistered(targetGroup, baragonAgents, elbsForBaragonGroup);
+
+            if (configuration.isPresent() && configuration.get().isDeregisterEnabled()) {
+              LOG.debug("De-registering old instances for group {}", baragonGroup.getName());
+              deregisterIfOld(baragonGroup, targetGroup, baragonAgents);
+            }
+          }
+
+          LOG.debug("ELB sync complete for group {}", baragonGroup);
+        }
+      }
+    } catch (AmazonClientException acexn) {
+      LOG.error("Could not retrieve elb information due to ELB client error {}", acexn);
+      exceptionNotifier.notify(acexn, ImmutableMap.of("groups", baragonGroups == null ? "" : baragonGroups.toString()));
+    } catch (Exception exn) {
+      LOG.error("Could not process ELB sync due to error {}", exn);
+      exceptionNotifier.notify(exn, ImmutableMap.of("groups", baragonGroups == null ? "" : baragonGroups.toString()));
+    }
+  }
+
+  private Collection<LoadBalancer> getLoadBalancersByBaragonGroup(BaragonGroup baragonGroup) {
+    Set<TrafficSource> trafficSources = baragonGroup.getSources();
+    Set<String> trafficSourceNames = new HashSet<>();
+    for (TrafficSource trafficSource : trafficSources) {
+      if (trafficSource.getType() == TrafficSourceType.APPLICATION) {
+        trafficSourceNames.add(trafficSource.getName());
+      }
+    }
+
+    DescribeLoadBalancersRequest loadBalancersRequest = new DescribeLoadBalancersRequest()
+        .withNames(trafficSourceNames);
+    return elbClient.describeLoadBalancers(loadBalancersRequest)
+        .getLoadBalancers();
+  }
+
+  /**
+   * Ensure that the given baragon agent is attached to the given load balancer. When this function
+   * completes, the baragon agent will be attached to the load balancer, whether or not it originally
+   * was.
+   *
+   * @param baragonAgents BaragonAgent to register with given load balancer
+   * @param loadBalancers Load balancer to register with
+   */
+  private void guaranteeRegistered(TargetGroup targetGroup,
+                                   Collection<BaragonAgentMetadata> baragonAgents,
+                                   Collection<LoadBalancer> loadBalancers) {
+    /*
+    - Check that load balancers, baragon agents, target groups are on same VPC
+    - Check that load balancers, targets are on same subnet (== AZ)
+    - Check that all baragon agents are associated with a target on target group
+    - Check that load balancers has listeners, rules to make talk to target group
+     */
+
+    if (configuration.isPresent() && configuration.get().isCheckForCorrectVpc()) {
+      guaranteeSameVPC(targetGroup, baragonAgents, loadBalancers);
+    }
+
+    guaranteeHasAllTargets(targetGroup, baragonAgents);
+    guaranteeAzEnabled(baragonAgents, loadBalancers);
```

Comment:
The az enable will need to go before the target addition most likely. In the case of the classic load balancers, if the az/subnet were not enabled, any registration in that az/subnet would fail. I would suspect it is likely the case here as well, although I can't find any docs that say so